Import transcript abundances from kallisto using tximport
================
Chelsea Herdman
February 25th, 2021

In order to perform differential expression analysis using DESeq2 on the
estimated transcript abundances produced by kallisto, we used the
tximport package in R.

We followed the vignette provided by Michael I. Love, Charlotte Soneson
and Mark D. Robinson found
[here](http://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html#salmon_with_inferential_replicates)

#### Prepare transcript to gene dataframe for gene-level summarization

Load required libraries.

``` r
library(tximport)
library(data.table)
library(ensembldb)
library(RMariaDB)
library(here)
library(rhdf5)
```

Fetch gene annotations from Ensembl and create one to one transcript
id/gene id data frame.

``` r
txdb = makeTxDbFromEnsembl(organism="Danio rerio",
                           release=89,
                           server="ensembldb.ensembl.org",
                           username="anonymous", password=NULL, port=0L,
                           tx_attrib=NULL)
```

    ## Fetch transcripts and genes from Ensembl ... OK
    ##   (fetched 58549 transcripts from 32266 genes)
    ## Fetch exons and CDS from Ensembl ... OK
    ## Fetch chromosome names and lengths from Ensembl ...OK
    ## Gather the metadata ... OK
    ## Make the TxDb object ... OK

``` r
k <- keys(txdb, keytype = "TXNAME")
tx2gene <- select(txdb, k, "GENEID", "TXNAME")
```

    ## 'select()' returned 1:1 mapping between keys and columns

``` r
head(tx2gene)
```

    ##               TXNAME             GENEID
    ## 1 ENSDART00000166393 ENSDARG00000102407
    ## 2 ENSDART00000170165 ENSDARG00000102407
    ## 3 ENSDART00000161842 ENSDARG00000102407
    ## 4 ENSDART00000170877 ENSDARG00000102407
    ## 5 ENSDART00000165461 ENSDARG00000102407
    ## 6 ENSDART00000171631 ENSDARG00000102407

``` r
length(unique(tx2gene$TXNAME))
```

    ## [1] 58549

``` r
length(unique(tx2gene$GENEID))
```

    ## [1] 32266

#### Prepare named vector of files

Use kallisto abundance.tsv files. These are listed in alphabetical order
in the kallisto\_out directory so ensure naming associates accurately
and that names associate with sample info table that will be used for
DESeq2.

``` r
file_path_vec = list.files(path=here("kallisto"),
                   pattern="abundance.tsv",
                   recursive = TRUE,
                   full.names = TRUE)
all(file.exists(file_path_vec))
```

    ## [1] TRUE

``` r
file_tab = data.table(file_path=file_path_vec,
                      sample_id=basename(dirname(file_path_vec)))

files = file_tab$file_path
names(files) = file_tab$sample_id
```

#### Run tximport for gene level estimation

##### Bias corrected counts without an offset

``` r
txi.kallisto.biascorr = tximport(files, type= "kallisto", tx2gene = tx2gene, 
                                 ignoreTxVersion = TRUE, 
                                 countsFromAbundance="lengthScaledTPM")
```

    ## Note: importing `abundance.h5` is typically faster than `abundance.tsv`

    ## reading in files with read.delim (install 'readr' package for speed up)

    ## 1 2 3 4 5 6 7 8 9 
    ## summarizing abundance
    ## summarizing counts
    ## summarizing length
    ## summarizing inferential replicates

``` r
txi.kallisto.biascorr$counts[1:6, 1:6]
```

    ##                    adult_heart_0 adult_heart_1 adult_heart_2 adult_muscle_0
    ## ENSDARG00000000001      118.0380      157.2532      171.3796       118.1167
    ## ENSDARG00000000002     1474.0669     1181.1402     1487.4960       378.6678
    ## ENSDARG00000000018      354.3242      552.3193      735.4752       491.5478
    ## ENSDARG00000000019     3426.5993     3292.2622     3919.2062      3260.6739
    ## ENSDARG00000000068      252.4972      242.8017      310.1968       390.3633
    ## ENSDARG00000000069      676.6983      627.0658      768.7963       393.3591
    ##                    adult_muscle_1 adult_muscle_2
    ## ENSDARG00000000001       71.77519       73.54006
    ## ENSDARG00000000002      269.52963      207.49239
    ## ENSDARG00000000018      491.93282      421.92992
    ## ENSDARG00000000019     2909.84116     2259.94184
    ## ENSDARG00000000068      401.14978      311.08177
    ## ENSDARG00000000069      419.61908      324.50507

``` r
counts_tab = as.data.table(txi.kallisto.biascorr$counts)
counts_tab$ensembl_gene_id = rownames(txi.kallisto.biascorr$counts)
setcolorder(counts_tab, c("embryonic_heart_0", "embryonic_heart_1", "embryonic_heart_2",
                          "adult_heart_0", "adult_heart_1", "adult_heart_2",
                          "adult_muscle_0", "adult_muscle_1", "adult_muscle_2"))
```

Save the counts table in order to use as input for DESeq2 (*see DESeq2
folder*)

``` r
fwrite(counts_tab, file=here("Tximport", "20210225_Wangetal_counts_fromtximport_biascorrected.txt"), sep="\t")
```
